syntax = "proto3";
import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";

package P2PFileSync.Protocol;

enum ProtoPayloadType{
  HELLO_MESSAGE = 0;
  REDIRECT_MESSAGE = 1;
}

message ProtoMessage {
  // sender/receiver information
  string sender_id = 5;
  string receiver_id = 6;

  // packet information
  string packet_id = 4;
  ProtoPayloadType request_type = 1;
  google.protobuf.Timestamp timestamp = 3;

  // actual payload
  google.protobuf.Any payload = 2;
}

// used for singed header
message ProtoSignedMessage{
  bytes signature = 1;
  google.protobuf.Any message = 2;
}

message ProtoHelloMessage {
  bytes public_key = 1;
}

message ProtoRedirectMessage {
  string orig_packet_id = 4;
  ProtoPayloadType orig_request_type = 1;
  google.protobuf.Any orig_payload = 2;
  google.protobuf.Timestamp orig_timestamp = 3;
}